### PR TITLE
Glimmer component

### PR DESCRIPTION
**Pull request by wycats**
Don't merge this yet, until the following items are complete:
- [x] reintroduce mergeAttrs & normalization for identity elements (`<my-foo>` top level inside `<my-foo>`) - done in d280c5353c3031272a3fda59e619e1d8f6fc6e02
- [x] remove attributes hook (post-processing for top-level elements) - done in d280c5353c3031272a3fda59e619e1d8f6fc6e02
- [x] figure out why `component.element` may not be set on `GlimmerComponent` fa01e2424d1e699c43abfaed37266339936c3305
- [x] identify and fix scope bug in identity elements c745c4766c371a3495049a9769876a6680ddbe85
- [x] make regular elements like identity elements -- done in ca436a2b1d4d488a6aeadd9ba7e6ab170e3a5e49 and 87e021a83559001dec692e6872a69ce2752f8f59
- [x] align fallback semantics with htmlbars (`<input value={{bar}} />`)
- [x] Re-enable previously skipped tests related to angle bracket components.
- [x] Fix and incorporate currently failing test from https://github.com/emberjs/ember.js/pull/11991 (test added in 6f359d2, fixed previously in c745c47)
- [x] Make sure `{{#with}}<div {{action}}></div>{{/with}}` works (64247979b8d112de7a89c259ca3e6edc9061abca)
- [x] Error when you use a fragment inside `GlimmerComponent` (5b686d44e3cadb174ed30a2f00e605eca1ebf37f)
  - [x] Better error when fragment comes from `{{action "foo"}}` or `style={{{bar}}}`
- [x] Make sure existing uses of already-registered web components continue to work (974493f0ddef29f607498d385d3ff3258ba00894)
- [x] all green??????????

Post-merge:
- [ ] Fix top-level `{{partial}}` case (0604db16ea85496b294d31f4289fbf1687924e89)
- [ ] Make sure that people don't accidentally use old-style APIs for top-level elements in GlimmerComponent (`tagName`)
- [ ] Refactor to separate conceptually different code paths that happen to share code, such as...
  - [ ] build-component-template
  - [ ] component node manager?
- [ ] make recursive invocation work (`<my-foo>` component as the top-level element inside `<my-bar>`'s layout)
- [ ] discuss a path for properties vs. attributes being plausibly aligned between normal elements and components

### CODE REVIEW

File: packages/ember-htmlbars/lib/hooks/bind-self.js [hunk 1]
```diff
@@ -19,7 +19,13 @@ export default function bindSelf(env, scope, _self) {
   if (self && self.isView) {
     newStream(scope.locals, 'view', self, null);
     newStream(scope.locals, 'controller', scope.locals.view.getKey('controller'));
-    newStream(scope, 'self', scope.locals.view.getKey('context'), null, true);
+
+    if (self.isGlimmerComponent) {
```
Comment (stefanpenner) [inline packages/ember-htmlbars/lib/hooks/bind-self.js]: should we instead use polymorphism here? Something like `this._setupSelfStream()` and let the subclasses refine further?

File: packages/ember-htmlbars/lib/node-managers/component-node-manager.js [hunk 2]
```diff
@@ -301,13 +310,13 @@ ComponentNodeManager.prototype.destroy = function() {
 export function createComponent(_component, isAngleBracket, _props, renderNode, env, attrs = {}, proto = _component.proto()) {
   let props = assign({}, _props);
 
+  let snapshot = takeSnapshot(attrs);
```
Comment (stefanpenner) [inline packages/ember-htmlbars/lib/node-managers/component-node-manager.js]: curious why this was pulled out of the branch

**Comment by runspired** [comment]
:D

File: packages/ember-htmlbars/lib/node-managers/component-node-manager.js (hunk 2 above)
Comment (wycats) [inline packages/ember-htmlbars/lib/node-managers/component-node-manager.js]: 8dc7d58cd4c0f45cc5aaf79c3ccf03d10dd7445f

Comment (stefanpenner) [inline packages/ember-htmlbars/lib/node-managers/component-node-manager.js]: ah thanks...

(would be nice if GH should which commit a specific change in a diff came from)

**Comment by chancancode** [comment]
#11728

**Comment by rwjblue** [comment]
We should address the plan (or intentional avoidance) of the nested component case (components in sub-dirs) https://github.com/emberjs/ember.js/issues/10456.

**Comment by wycats** [comment]
@rwjblue seems legit. I assume you mean via documentation, right?

**Comment by rwjblue** [comment]
@wycats - Yeah, we should have docs and tests that say exactly what we do and don't support subdirectory wise.  I'm not suggesting that we have to support subdirs for landing + enabling, but I'd like to be very clear about what we intend (and that we didn't "forget").

**Comment by chancancode** [comment]
We also need to double check the GlimmerComponent class is exported correctly (right module path and globally mode too?)

**Comment by chadhietala** [comment]
Probably needs an RFC, but it would be great if glimmer components could optionally specified required types of attrs during dev.  It would make integrating components from addons a much more pleasant experience, might also have perf wins in prod due to the fact that the code was designed to be called consistently with the same object types. This would be like `propTypes` in react.

**Comment by runspired** [comment]
@chadhietala +1 on that, I try to write helpful asserts but also hate writing them all the time.

**Pull request merged**